python_modules/dagster/dagster/_core/execution/context/resource_invocation.py

```diff
@@ -0,0 +1,112 @@
+        else:
+            self._resources = None
+            self._resource_defs = wrap_resources_for_execution(
+                check.opt_mapping_param(resources, "resources", key_type=str)
+            )
+            self._resources_config = check.opt_mapping_param(resources_config, "resources_config")
+
+    def get_resources(self) -> Resources:
+        from dagster._core.execution.build_resources import build_resources
+
+        if self._resources is None:
+            with build_resources(
+                resources=self._resource_defs,
+                instance=self._instance,
+                resource_config=self._resources_config,
+            ) as resources:
+                # We only find out that there is a generator resource after
+                # initialization, but ideally, we should be able to find out
+                # before resource initialization occurs.
+                if isinstance(resources, IContainsGenerator):
+                    raise DagsterInvariantViolationError(
+                        "At least one provided resource is a generator, but attempting to access "
+                        "resources outside of context manager scope. You can use the following syntax to "
+                        f"open a context manager: `with build_{self._context_str}_context(...) as context:`"
+                    )
+                self._resources = resources
+        return self._resources
+
+    def get_resource_defs(self) -> Mapping[str, ResourceDefinition]:
+        if self._resource_defs is None:
+            raise DagsterInvariantViolationError(
+                "Resource defs were not directly provided to this context."
+            )
+        return cast(Mapping[str, ResourceDefinition], self._resource_defs)
+
+    def get_resources_config(self) -> Mapping[str, Any]:
+        if self._resources_config is None:
+            raise DagsterInvariantViolationError(
+                "Resources config was not directly provided to this context."
+            )
+        return cast(Mapping[str, Any], self._resources_config)
+
+    def __enter__(self):
+        from dagster._core.execution.build_resources import build_resources
+        from dagster._core.execution.api import ephemeral_instance_if_missing
+
+        if not self._instance:
+            self._instance_cm = ephemeral_instance_if_missing(None)
+            self._instance = self._instance_cm.__enter__()
+
+        self._resources_cm = build_resources(
+            resources=self._resource_defs,
+            instance=self._instance,
+            resource_config=self._resources_config,
+        )
+        self._resources = self._resources_cm.__enter__()
+        return self
+
+    def __exit__(self, *exc):
+        if self._instance_cm:
+            self._instance = None
+            self._instance_cm.__exit__(*exc)
+        if self._resources_cm:
+            self._resources = None
+            self._resources_cm.__exit__(*exc)
```

Comment:
when do we still need exit clean up?

Reply:
Some resources will require cleanup (eg a finally statement). For resources that have this, we need to make sure that cleanup doesn't occur until after the context manager ie `with build_x_context` is exited. And we only want to call exit if the cm actually exists.

Hmm I guess I'm thinking now that we also need to check whether instance_cm / resources_cm has already been exited, so maybe we set those to none after exiting as well